### ISSUE TITLE
GLES3 needs manual color correction, GLES2 does not (and produces the wrong one if corrected)

### ISSUE THREAD

**Issue by rokasv** [godotengine/godot#25142]
Windows 10
Godot 3.1.beta2
GeForce GTX 1050 Ti/PCIe/SSE2

GLES3 and GLES2 produce significantly different colors when drawn from viewport textures.

Scene tree:
![tree](https://user-images.githubusercontent.com/13801432/51430289-10991380-1c11-11e9-8dbb-1f242ae6299b.PNG)

TextureRect texture:
<img src="https://user-images.githubusercontent.com/13801432/51430311-666dbb80-1c11-11e9-9bd8-377b13d7e2f2.png" width="200" height="200">

GLES2 output (correct):
![gles2_vp](https://user-images.githubusercontent.com/13801432/51430293-23abe380-1c11-11e9-84af-c80e424dd9f0.PNG)

GLES3 output (wrong):
![gles3_vp](https://user-images.githubusercontent.com/13801432/51430298-2c9cb500-1c11-11e9-8d4e-4a740d3e5e60.PNG)

The contrast inconsistencies break certain visuals, making one mode unusable.
Reproduce with [project.zip](https://github.com/godotengine/godot/files/2775815/project.zip)

**Comment by Calinou** [comment]
This can be fixed in GLES3 (without impacting the appearance in GLES2) by importing the texture as sRGB. You can do this by selecting the colors.png image in the FileSystem dock, enabling sRGB then reimporting it like below:

![image](https://user-images.githubusercontent.com/180032/51430487-50192d00-1c1c-11e9-8b62-0ba8bb77d781.png)

**Comment by rokasv** [comment]
I think I know the issue:

This is the code used by SpatialShaders to ensure the final screen color is correct:
```glsl
vec3 adjust_rgb(vec3 color){
	return mix(pow((color + vec3(0.055)) * (1.0 / (1.0 + 0.055)),vec3(2.4)),color * (1.0 / 12.92),lessThan(color,vec3(0.04045)));
}
```

When using my own shaders I include it too. But the code only appears to be necessary in GLES3. 
When included, the color is correct in GLES3, but not GLES2. 
When removed, the color is correct in GLES2, but not GLES3.

Specifically:
GLES3 adjusted output (correct):
![gles3_adjusted](https://user-images.githubusercontent.com/13801432/51432017-01be5b00-1c29-11e9-90ce-13929b004bc1.PNG)

GLES3 not adjusted (wrong):
![gles3_not_adjusted](https://user-images.githubusercontent.com/13801432/51432022-0daa1d00-1c29-11e9-9dc1-f4c4ef9ea8c6.PNG)

GLES2 adjusted (wrong):
![gles2_adjusted](https://user-images.githubusercontent.com/13801432/51432029-1f8bc000-1c29-11e9-831b-48a7f3857e3d.PNG)

GLES2 not adjusted (correct):
![gles2_not_adjusted](https://user-images.githubusercontent.com/13801432/51432032-29adbe80-1c29-11e9-8838-77063d77655d.PNG)

I'm guessing the `adjust_rgb()` shadercode is `IFDEF`ed in the default engine shader.

In summary: the color is automatically corrected in GLES2, but needs to be manually corrected in GLES3, if using custom shaders.

--------------------
Until a patch is issued, is there a way to detect which render is running from GDScript, so I can select the appropriate shaders?

**Comment by rokasv** [comment]
Here's the [project.zip](https://github.com/godotengine/godot/files/2775995/project.zip) to reproduce it.

Verification:
1. Run project in GLES3
2. Toggle this bool on the root node
![tree](https://user-images.githubusercontent.com/13801432/51432566-a775c800-1c31-11e9-88bf-62ba87ff0f54.PNG)
3. Compare the colors with the ones below
4. Run project in GLES2
5. Toggle the bool and compare the colors with the ones below
6. Notice that GLES3 adjusted and GLES2 not adjusted are identical and correct

GLES2 adjusted (wrong):
![gles2_adjusted](https://user-images.githubusercontent.com/13801432/51432579-cd02d180-1c31-11e9-9e42-233b3288ec7d.PNG)

GLES2 not adjusted (correct):
![gles2_not_adjusted](https://user-images.githubusercontent.com/13801432/51432583-dbe98400-1c31-11e9-835a-f09b2a7562aa.PNG)

GLES3 adjusted (correct):
![gles3_adjusted](https://user-images.githubusercontent.com/13801432/51432591-e73caf80-1c31-11e9-9564-85bdf5dd9279.PNG)

GLES3 not adjusted (wrong):
![gles3_not_adjusted](https://user-images.githubusercontent.com/13801432/51432594-f02d8100-1c31-11e9-927e-c8d77bdd41e6.PNG)

**Comment by Calinou** [comment]
> Until a patch is issued, is there a way to detect which render is running from GDScript, so I can select the appropriate shaders?

You could use this, which will return `GLES3` or `GLES2`:

```gdscript
ProjectSettings.get("rendering/quality/driver/driver_name")
```

However, it will still return `GLES3` if it falls back automatically to GLES2 due to unsupported hardware.

**Comment by rokasv** [comment]
Just to make sure, is this considered a bug?

I'm not sure if I should be reporting GLES version inconsistencies which can be worked around, since you could technically just use GLES2.

**Comment by clayjohn** [comment]
Try using 
```
if (!OUTPUT_IS_SRGB) {
    vec3 adjust_rgb(vec3 color){
	    return mix(pow((color + vec3(0.055)) * (1.0 / (1.0 + 0.055)),vec3(2.4)),color * (1.0 / 12.92),lessThan(color,vec3(0.04045)));
    }
}
```
GLES2 and GLES3 handle sRGB differently. I am not sure exactly why. But this is the reason for the ``OUTPUT_IS_SRGB`` flag.

**Comment by rokasv** [comment]
`OUTPUT_IS_SRGB` is not accessible from the Godot shading language.
 
This type of solution would be good enough.

**Comment by rokasv** [comment]
The `OUTPUT_IS_SRGB` does it, thanks, though not for defining functions. i.e.
```
vec3 color = vec3(1.0, 1.0, 0.1);
if (!OUTPUT_IS_SRGB) {
	color = adjust_rgb(color);
}
```
Including this logic in the default shader sounds like a good idea to me.

**Comment by clayjohn** [comment]
Ideally they would both work in the same color space. I'm not sure why they currently don't.

**Comment by akien-mga** [comment]
For the time being this probably needs to be better documented.

**Comment by clayjohn** [comment]
I have added this to a tracker for differences between GLES2 and GLES3. Ideally we will put together a document before 3.1 releases.

**Comment by clayjohn** [comment]
Now that this has been appropriately documented we can close the issue.
